src/geopackage/sqlite/iterator.cpp

```diff
@@ -0,0 +1,152 @@
+#include <algorithm>
+
+#include "NGen_SQLite.hpp"
+
+using namespace geopackage;
+
+// Defined in database.cpp
+extern std::runtime_error sqlite_error(const std::string& f, int code, const std::string& extra = "");
+
+/**
+ * Runtime error for iterations that haven't started
+ */
+const auto sqlite_get_notstarted_error = std::runtime_error(
+    "sqlite iteration is has not started, get() is not callable (call sqlite_iter::next() before)"
+);
+
+/**
+ * Runtime error for iterations that are finished
+ */
+const auto sqlite_get_done_error = std::runtime_error(
+    "sqlite iteration is done, get() is not callable"
+);
+
+sqlite_iter::sqlite_iter(stmt_t stmt)
+  : stmt(stmt)
+{
+    // sqlite3_column_type requires the last result code to be
+    // SQLITE_ROW, so we need to iterate on the first row.
+    // TODO: need to test how this functions if the last result code
+    //       was SQLITE_DONE.
+    this->next();
+    this->column_count = sqlite3_column_count(this->ptr());
+    this->column_names = std::vector<std::string>();
+    this->column_names.reserve(this->column_count);
+    this->column_types = std::vector<int>();
+    this->column_types.reserve(this->column_count);
+
+    for (int i = 0; i < this->column_count; i++) {
+        this->column_names.emplace_back(sqlite3_column_name(this->ptr(), i));
+        this->column_types.emplace_back(sqlite3_column_type(this->ptr(), i));
+    }
+
+    this->restart();
+};
+
+sqlite3_stmt* sqlite_iter::ptr() const noexcept
+{
+    return this->stmt.get();
+}
+
+void sqlite_iter::handle_get_index(int col) const
+{
+    if (this->done()) {
+        throw sqlite_get_done_error;
+    }
+
+    if (this->current_row() == -1) {
+        throw sqlite_get_notstarted_error;
+    }
+
+    if (col < 0 || col >= this->column_count) {
+        
+        throw std::out_of_range(
+            "column " + std::to_string(col) + " out of range of " + std::to_string(this->column_count) + " columns"
+        );
+    }
+}
+
+bool sqlite_iter::done() const noexcept
+{
+    return this->iteration_finished;
+}
+
+sqlite_iter& sqlite_iter::next()
+{
+    if (!this->done()) {
+        const int returncode = sqlite3_step(this->ptr());
+        if (returncode == SQLITE_DONE) {
+            this->iteration_finished = true;
+        } else if (returncode != SQLITE_ROW) {
+            throw sqlite_error("sqlite3_step", returncode);
+        }
+        this->iteration_step++;
+    }
+
+    return *this;
+}
+
+sqlite_iter& sqlite_iter::restart()
+{
+    sqlite3_reset(this->ptr());
+    this->iteration_step = -1;
+    this->iteration_finished = false;
+    return *this;
+}
+
+void sqlite_iter::close()
+{
+    this->~sqlite_iter();
```

Comment:
Writing an explicit call to a class's own destructor is a code smell. This is primarily a round-about way of trying to say `sqlite3_finalize(ptr())`, right? It doesn't even really do that, because the `sqlite` instance we go it from still has a reference to it.

Reply:
Implemented changes in 914ec44: removed `sqlite_iter::close()`, this function wasn't being used anywhere, so I'm not sure why I implemented it (or why I implemented it to call the destructor) -- might've been old code